Add CTE receive and sampling in Bluetooth stack

ppryga-nordic opened the pull request:
The PR provides CTE receive and sampling functionality in BLE stack. It includes all three parts of implementation: controller, HCI, host.

The controller part: 
- Added changes in radio implementation to enable CTE reception for periodic advertising scan events.
- Added API to get information about antenna id used to measure RSSI.
- Added API to get CTEInfo PDU field parsed by radio peripheral.
- New noder_rx type introduced. Thanks to that we are able to manage memory for required collected IQ samples separately from memory for regular PDUs. The node_rx_iq_report has node_rx_hdr, so it may be processed by regular received nodes path up to HCI.
- Added enable or disable CTE receive in LL thread and ULL. The CTE enable and sampling configuration are provided to LLL via double buffer (stored in lll_sync). That functionality works in a similar way like double buffered PDUs in advertising role.
- Added handling of HCI_LE_Set_Connectionless_IQ_Sampling_Enable HCI command.
- Added dropping of unreported node_rx_iq_repots if CTE receive was disabled by Host in the meantime (e.g. disable operation was preempted by periodic advertising sync event and new node_rx_iq_report was put in receive path).
- Added unit tests for handling of HCI_LE_Set_Connectionless_IQ_Sampling_Enable.

The HCI part:
- Added HCI_LE_Set_Connectionless_IQ_Sampling_Enable command.
- Added HCI_LE_Connectionless_IQ_Report.

The host part:
- Added API to enable or disable CTE receive and sampling.
- Added API to register callbacks for IQ samples report receive.
- Added handling of HCI_LE_Connectionless_IQ_Report event.

The PR is long. I'm sorry for that. Splitting it at this stage (in my humble opinion) would increase the time it may be merged to master.

ppryga-nordic committed:
Bluetooth: controller: New node_rx type for IQ sample report

Add new kind of node_rx type that will store IQ samples report.
The structure includes node_rx_hdr and may be used by the same
queues that are used to forward regular node_rx_pdu objects
between LLL and ULL.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add IQ report mem pool management

Add functions that will allow to mangage IQ reports memory pool
e.g. allocate or relase new IQ report nodes.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Disable not implemented DF features

Set to disabled not yet implemented Direction Finding features.
The features will be enabled by default when they are implemented.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add DF per. adv. sync. configuration stucture

Add new structure lll_df_sync_cfg to store DF configuration
for periodic advertising synchronization.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add double buffer for DF sync configuration

CTE sampling enable or disable requests should be provided to LLL.
Double buffer of lll_df_sync_cfg is added to lll_sync for that purpose
Thanks to that:
- there is no need to synchronize LLL with LL thread
- LLL has always access to valid configuration (new or former)

The configuration is stored in double buffer as instances
of lll_df_sync_cfg, not as pointers. This is the only
difference between double buffer pattern used to store adverising PDU.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
include: Bluetooth: hci: move command to appropriate position in file

All commands are put in the file according to OpCode Command Field
increasing order. bt_hci_cp_le_set_cl_cte_tx_enable was written in
wrong position. This commit moves it to correct one.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
include: Bluetooth: hci: Add LE set connectionless IQ sampling enabled

Add new command HCI_LE_Set_Connectionless_IQ_Sampling_Enable that
will allow host to enable CTE sampling in controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
include: Bluetooth: hci: Add LE connectionelss IQ samples report event

Add HCI_LE_Connectionless_IQ_Report event that will allow controller
to report received CTE and provide collected IQ samples to host.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add Kconfig option to enable debugging code in DF

To avid enabling HCI drived debugging, like in other controller source
files, Direction Finding has its own option BT_CTLR_DF_DEBUG_ENABLE.
This gives fine grained control deubgging output.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: hci: Add connectionelss IQ samplig enable handler

Add handler for HCI_LE_Set_Connectionless_IQ_Sampling_Enable HCI
command.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: hci: Add connectionless IQ report event handler

Add handler for HCI_LE_Connection_IQ_Report event.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: radio: Add function to get PDU ant. switch pattern

IQ samples report requires information about antenna switch pattern
that was used to measure RSSI. RSSI is measured during PDU receive.
For Nordic Radio peripheral the PDU antenna is selected by first antenna
switch pattern set in SWITCHPATTERN register. That pattern is providded
by device tree. This information is already available in radio_df.c.
The information is required by LLL, that has access to functions
provided in radio_df.c. New function was provided to return PDU antenna
switch pattern.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Add node_rx_iq_report handling in ULL RX path

Add handling of Direction Finding node_rx_iq_report type in ULL.
This is required to correctly deliver IQ samples collected by
LLL during receive of PDU including CTE.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: Add CTE sampling and reporting

ULL may request LLL to start receiving and sampling CTE attached to
end of periodic advertising PDUs. When the CTE is successfully sampled,
LLL should report collected samples to ULL.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Add function to enable or disable CTE sampling

Add function that enables or disables CTE sampling attached to received
periodic advertising PDUs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: Add DF configuration to lll_sync structure.

To enable or disable CTE sampling in periodic advertising scanner LLL
implementation lll_sync is extended by lll_df_sync member.
The lll_df_sync is a double buffer for lll_df_sync_cfg.
It will be used by LL thread to enable or disable CTE sampling and
provide configuration for sampling CTE in AoA mode.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Add init of DF sync configuration

lll_sync object is used to store information for LLL to enable
or disable CTE sampling. The commit adds initialization of
lll_df_sync member of lll_sync.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: radio: Add function to get CTEInfo from Radio reg

Nordic Radio perihperal provides possiblity to get value of CTEInfo
field parsed from received PDU. It is faster to get the information
from Radio register than parse a PDU content.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add function to update per sync event slot duration

To enable or disable CTE sampling during periodic advertising sync evnet
it is requireq to change the event duration. To make it possible new
function was added that handles the operation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: DF: Add function to enable CTE receive

Add function to enable CTE receive and sampling. The function
is aimed to be used by code in lll_sync.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add Kconfig option to enable conn-less CTE sampling

Direction Finding support requires possibility to receive and sample CTE
attached to periodic advertising PDUs. New option will be used to enable
or disable the functionality. New option is introduced to avoid using
double statements checking if BT_CTLR_DF_CTE_RX and BT_CTLR_SYNC_-
-PERIODIC are enabled.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add opt. to enable CTE samplig for PDUs with bad CRC

Add Kconfig option that will enable possiblity to sample CTE attached to
PDUs that have bard CRC. Sampling is based on CTEInfo field available
in received PDU. If radio is able to parse the field correclty then
it will start sampling even the CRC is wrong.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: Add CTE report for per. adv. PDU with bad CRC

Add reporing of IQ samples collected for CTE attached to periodic
advertising PDUs with bad CRC.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Dropp IQ sampl report if CTE disabled after evt end

Disable of CTE sampling for periodic advertising sync PDUs may be
interrupted by radio event. In such situation, TX thread will be
handled before RX thread delivers IQ samples report to host. In
this case RX thread would deliver IQ samples report after CTE
sampling was disabled.

To avoid such situation RX thread has to check if:
- CTE sampling is not disabled,
- CTE sampling was not requested to be disabled after start of radio
  event.

CTE sampling configuration is double buffered. Updated configuration is
swapped at the beginning of radio event.

If CTE sampling is diabled or requested to disable (next radio event
didn't start), the IQ samples report should be dropped by RX thread
and should not be delivered to host.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Move UT code into connectionless_cte_tx sub dir

Before thic commit the bluetooth/df unit tests subdirectory contained
connectionless CTE TX verification. Due to the implementation of
connectionless CTE RX new set of tests will be provided.
The new functionality requires different: configuration and common
setup code. Because of that, current test code was moved to
connectionless_cte_tx subdirectory.
Code shared by all DF tests is moved to common subdirectory.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Add connectionless IQ sampling enable unit tests

Add unit tests for connectionless IQ sampling functionality in
controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add connectionless CTE RX feat. in supp. feat. list

Add connectionless CTE RX feature to list of features supported
by controller. Add direction finding initialization in hci_core
if the feature is enabled.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: df: Set BT_CTLR_DF_ANT_SWITCH_RX enable

Set BT_CTLR_DF_ANT_SWITCH_RX enabled by default.
There is already added implementation that supports
this feature in controller, so it may be enabled by
default.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: Add option to enable connectionless CTE RX

Add Kconfig option that will enable supprot for connectionless
CTE reception. Thanks to that it will be possible to conditionally
enable or disable support of the feature and decrease code size
if the feature is not required.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: Add CTE receive and sample enable API

Add functions that give possibility to enable or disable
CTE receive and sample in connectionless mode.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: Add CTE reports receive cb API for per sync

Add new callback to periodic advertising sync callback.
The callback may be used to receive notification about CTE
reports in connectionless mode (periodic advetising sync).

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: Add handl HCI_LE_Connectionless_IQ_Report

Add handling for HCI_LE_Connectionless_IQ_Report received
from controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
The PR provides CTE receive and sampling functionality in BLE stack. It includes all three parts of implementation: controller, HCI, host.

The controller part: 
- Added changes in radio implementation to enable CTE reception for periodic advertising scan events.
- Added API to get information about antenna id used to measure RSSI.
- Added API to get CTEInfo PDU field parsed by radio peripheral.
- New noder_rx type introduced. Thanks to that we are able to manage memory for required collected IQ samples separately from memory for regular PDUs. The node_rx_iq_report has node_rx_hdr, so it may be processed by regular received nodes path up to HCI.
- Added enable or disable CTE receive in LL thread and ULL. The CTE enable and sampling configuration are provided to LLL via double buffer (stored in lll_sync). That functionality works in a similar way like double buffered PDUs in advertising role.
- Added handling of HCI_LE_Set_Connectionless_IQ_Sampling_Enable HCI command.
- Added dropping of unreported node_rx_iq_repots if CTE receive was disabled by Host in the meantime (e.g. disable operation was preempted by periodic advertising sync event and new node_rx_iq_report was put in receive path).
- Added unit tests for handling of HCI_LE_Set_Connectionless_IQ_Sampling_Enable.

The HCI part:
- Added HCI_LE_Set_Connectionless_IQ_Sampling_Enable command.
- Added HCI_LE_Connectionless_IQ_Report.

The host part:
- Added API to enable or disable CTE receive and sampling.
- Added API to register callbacks for IQ samples report receive.
- Added handling of HCI_LE_Connectionless_IQ_Report event.

The PR is long. I'm sorry for that. Splitting it at this stage (in my humble opinion) would increase the time it may be merged to master.

ppryga-nordic committed:
Bluetoot: host: Make get_per_adv_sync to be global function

The function get_per_adv_sync is used by HCI events to get
reference to periodic advertising sync object related with
handle available in handled events. Due to implementation
of Direction Finding event handlers in separate source file,
direction.c instead of putting in hci_core.c the function
get_per_adv_sync has to be globally accessible.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Resolved merge conflicts.

ppryga commented:
Updated commit messages.

ppryga-nordic committed:
Bluetooth: controller: Fix build error for non Nordic plaforms

Direction finding is implemented only for Nordic link layer.
Build was failing due to missing header files for Openisa (RiskV).
Added dummy lll_df_types.h header file to OpenISA lower link layer.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Fixed issues in unit tests. Added commit that fixes builds of controller for Openisa (RiscV) platforms.

ppryga commented:
Rebased due to failing clang/llvm builds.

ppryga-nordic committed:
samples: Bluetooth: Add Direction Finding connectionless Rx sample

Add an application that uses Direction Finding API for reception
and sampling of CTE in connectionless mode (periodic adverising PDUs).

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Changes implementation of `hci_df_set_cl_cte_rx_enable` to hide an issue with Bluetooth Core Spec. Vol 4,Part E, sec 7.8.82. There is no right way to set receive of AoD CTE for devices with single antenna. The spec requires minimum antenna switch pattern to have at least two antennas. 

 To hide that from end user, the function verifies cte_type. If there is an AoA type enabled, slot_durations, num_ant_ids and ant_ids must contain valid values. 

Besides that cte_type is stored in `struct bt_le_per_adv_sync`. `hci_le_df_connectionless_iq_report` uses it to drop IQ sample reports that has type not set in cte_type.

ppryga commented:
Added commit with sample application for CTE receive and sampling.

ppryga commented:
Before merge of this PR, that must be merged: https://github.com/zephyrproject-rtos/zephyr/pull/32358

ppryga commented:
@joerchan @jhedberg could you take a look on host part of the PR?
@nashif could you take a look on sample application?

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
 	uint8_t  *ant_ids;
 };
 
+enum BT_DF_CTE_RX_TYPE {
+	BT_DF_CTE_RX_TYPE_AOA = BIT(0),
+	BT_DF_CTE_RX_TYPE_AOD_1US = BIT(1),
+	BT_DF_CTE_RX_TYPE_AOD_2US = BIT(2),
+};
```
Does this differ from the GAP enum? 
```
/** Constant Tone Extension (CTE) types */
enum {
	/** Angle of Arrival */
	BT_GAP_CTE_AOA = 0x00,
	/** Angle of Departure with 1 us slots */
	BT_GAP_CTE_AOD_1US = 0x01,
	/** Angle of Departure with 2 us slots */
	BT_GAP_CTE_AOD_2US = 0x02,
	/** No extensions */
	BT_GAP_CTE_NONE = 0xFF,
};
```

ppryga replied:
The `enum BT_DF_CTE_RX_TYPE` members values are bit field values. They are used to set `bt_df_per_adv_sync_cte_rx_param::cte_type`. Why bit field, because an application may expect to receive IQ samples report for CTE transmitted in AoA and AoD(1us, 2us) modes. At least BT 5.1 spec allows for that.

ppryga replied:
Enum defined in `gap.h` seems to be not used anywhere now. What more, in my humble opinion, it belongs more to direction finding than to GAP. 
`HCI_LE_Periodic_Advertising_Create_Sync` command, that seems to be related with enum from gap.h, also uses bitfield for `sync_cte_type` parameter. Actually `sync_cte_type` has more allowed values, `BT_DF_CTE_RX_TYPE` is a subset of those. @Thalley correct me if I'm wrong, please.

ppryga replied:
I've given a thought to @joerchan proposal to unify both enums. I suggest is: don't do it. 
Enum provided in gap.h has different purpose and there is missing one value: `BT_GAP_CTE_ONLY`. The purpose of the enum seems to be limiting a periodic advertising syncs depending of CTE type.

The CTE types used by Direction finding RX and TX are subset of the CTE type defined in gap.h. But pay attention that DF does not use something like `CTE_NONE` or `CTE_ONLY` and that may be confusing for app developers.

I will unify CTE types but only within DF implementation. I'll create an issue in github to update DF connectionless TX parameters to use the unified version of CTE types after this PR is merged.

@Thalley @joerchan what do you think about it?

Thalley replied:
The GAP enum is used for the `struct bt_le_per_adv_sync_recv_info` at least, among others (or at least that was the intention). 

I understand why you argue for it to be in direction finding, which would be the only implemented usecase of it (and as far as I know, the only SIG defined use case), but nonetheless it is a generic feature that may be used in the application in other ways. I had originally defined them in bluetooth.h but moved them to gap.h on @joerchan's request. 

Regardless, I believe that it makes the most sense of having a single definition for these CTE types for the public API.  How they are transformed/used towards HCI can be dealt with in other ways. Our goal should be to keep the public API simple.

ppryga replied:
Ok, I'll create a common enum that will be used in this PR and also for `struct bt_le_per_adv_sync_recv_info`. The enum will be a bit-field, because I want to use it also for RX part of DF where more than one type of CTE may be expected. 
That will not prevent use of the enum values as integers for CTE type in DF TX and periodic advertising reports. 
Update of DF TX and periodic advertising will be done in separate PR. Then I'll also remove the enum from gap.h.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+struct bt_df_per_adv_sync_cte_rx_param {
+	/* Bitmap with allowed CTE types: AoD 1us, AoD 2us, AoA */
+	uint8_t cte_type;
+	/** Antenna switching slots, one of possible values: 1us, 2us. */
+	uint8_t slot_durations;
```
How does the 1us and 2us here compare to the AoD 1us and AoD 2us?

ppryga replied:
These values (1us and 2us) are related with AoA. In case `cte_type` has bit BT_DF_CTE_RX_TYPE_AOA unset. Then those values will be not verified in `bt_df_per_adv_sync_cte_rx_enable` call.

Thalley replied:
Hmm, I'm just wondering if it's clear to the applicaiton what the possible values can actually be here. Since it's one of the two values, it doesn't make sense to expect values of the aforementioned enum. Is is such that 1 = 1us and 2 = 2us, or is it 0 = 1us, and 1 = 2us?

ppryga replied:
I've created an enum with allowed values. I'll put a reference to the enum here.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	/** Max number of CTEs to receive, 0 means receive continuously. */
+	uint8_t max_cte_count;
+	/** Length of antenna switch pattern */
+	uint8_t num_ant_ids;
+	/** Antenna switch pattern */
+	const uint8_t *ant_ids;
```
What is the maximum number of antenna IDs?

ppryga replied:
Maximum number of allowed antenna IDs is 75, That is a value provided by BT 5.1. spec. But in case of controller implementation for Nordic SOCs the value is limited to 39. Actual max value is read by hci_df_read_ant_info call to controller. Returned by the function max switch pattern length holds CONFIG_BT_CTLR_DF_MAX_ANT_SW_PATTERN_LEN value.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	struct bt_hci_le_iq_sample const *sample;
+};
+
+struct bt_df_per_adv_sync_cte_report_cb {
+	/**
+	 * @brief Callback for IQ samples report collected when sampling
+	 *        CTE received with periodic advertising PDU.
+	 *
+	 * @param sync The periodic advertising sync object.
+	 * @param info Information about the sync event.
+	 */
+	void (*recv)(struct bt_le_per_adv_sync *sync,
+		     struct bt_df_per_adv_sync_iq_samples_report const *info);
+
+	sys_snode_t node;
+};
```
I'm wondering if it would be better to simple add this callback into the existing periodic advertising sync callback structure. Did you consider that?

ppryga replied:
I thought about it, but I decided to go with separate list for CTE receive callbacks because:
- wanted to have complete feature in single place
- DF uses periodic advertising sync as kind of transport for CTE
- CTE receive callback will not confuse users that do not know anything about DF and do not  want to use it.

@Thalley What is your thought about it?

ppryga replied:
If Broadcast ISO callback is in PA sync callback, I'll move the cte_report_cb there also.

ppryga commented:
I've found an issue in `lll_sync.c` `isr_rx`, handling of CTE receive while CRC was not ok scheduled ull unconditionally.

carlescufi commented:
> Before merge of this PR, that must be merged: #32358

This is merged now, please rebase.

ppryga commented:
> > Before merge of this PR, that must be merged: #32358
> 
> This is merged now, please rebase.

Done

carlescufi commented:
@Thalley @joerchan @cvinayak @kruithofa could you please review?

kruithofa reviewed (approved):
LGTM

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -102,6 +112,9 @@ static int init_reset(void)
 static int prepare_cb(struct lll_prepare_param *p)
 {
 	struct node_rx_pdu *node_rx;
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX)
+	struct lll_df_sync_cfg *cfg;
+#endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */
```
```suggestion
```

ppryga replied:
Moved `cfg` definition down, just before it is used in function flow.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -165,11 +178,29 @@ static int prepare_cb(struct lll_prepare_param *p)
 
 	lll_chan_set(data_chan_use);
 
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX)
+	cfg = lll_df_sync_cfg_latest_get(&lll->df_cfg, NULL);
+
+	if (cfg->is_enabled) {
+		lll_df_conf_cte_rx_enable(cfg->slot_durations,
+					  cfg->ant_sw_len,
+					  cfg->ant_ids);
+		cfg->cte_count = 0;
+	}
+#endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */
+
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX)
	struct lll_df_sync_cfg *cfg;

	cfg = lll_df_sync_cfg_latest_get(&lll->df_cfg, NULL);

	if (cfg->is_enabled) {
		lll_df_conf_cte_rx_enable(cfg->slot_durations,
					  cfg->ant_sw_len,
					  cfg->ant_ids);
		cfg->cte_count = 0;
	}
#endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */

```

ppryga replied:
Moved `cfg` as suggested.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -35,6 +35,9 @@
 #include "lll_adv.h"
 #include "lll/lll_adv_pdu.h"
 #include "lll_scan.h"
+#if defined(CONFIG_BT_CTLR_DF)
+#include "lll/lll_df_types.h"
+#endif /* CONFIG_BT_CTLR_DF */
```
```suggestion
#include "lll/lll_df_types.h"
```

ppryga replied:
Conditions around include removed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -45,7 +48,9 @@
 #include "ull_sync_types.h"
 #include "ull_conn_types.h"
 #include "ull_filter.h"
+#if defined(CONFIG_BT_CTLR_DF)
 #include "ull_df.h"
+#endif /* CONFIG_BT_CTLR_DF */
```
```suggestion
#include "ull_df.h"
```

ppryga replied:
Conditions around include removed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -804,7 +809,12 @@ void ll_rx_dequeue(void)
 	LL_ASSERT(link);
 
-	mem_release(link, &mem_link_rx.free);
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX)
+	if (rx->type != NODE_RX_TYPE_IQ_SAMPLE_REPORT)
+#endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */
+	{
+		mem_release(link, &mem_link_rx.free);
+	}
```
cross release of `link` buffer into pool. memq link supplied when enqueuing is used for next enqueue. Prefer the link come from same pool, unless this has been analysed.

ppryga replied:
The `node_rx_iq_report` link may not be stored in general `node_rx_pdu` memq link. `node_rx_iq_report` has different size than `node_rx_pdu`. Because of that there are separate memory pools and separate link memq for both types of `node_rx`.
Nevertheless I didn't want to create completely new data path from LLL to ULL thread, just to forward `node_rx_iq_report` nodes. It would introduce more complications than use of the same node forwarding mechanism. It might also introduce more code to be executed than this if statement. 

The if statement may be removed if release will be moved to particular case statements, but then there will be more code lines.

cvinayak replied:
If you move the content of `ull_df_rx_iq_report_alloc`  into `rx_alloc`, as you already are using a separate quota and separate free MFIFO, you only need the `ull_iq_report_link_inc_quota`. You can remove `ull_df_rx_iq_report_alloc`, `ull_df_iq_report_link_release` and increase `LINK_RX_POOL_SIZE`.

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/prj.conf:
```diff
@@ -0,0 +1,16 @@
+CONFIG_BT=y
+CONFIG_BT_CTLR=y
+CONFIG_BT_LL_SW_SPLIT=y
+CONFIG_BT_DEVICE_NAME="DF Connectionless Locator App"
+
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_PER_ADV_SYNC=y
+CONFIG_BT_OBSERVER=y
+CONFIG_BT_CTLR_ADV_EXT=y
+CONFIG_BT_CTLR_SYNC_PERIODIC=y
+
+# Enable Direction Finding Feature including AoA and AoD
+CONFIG_BT_DF=y
+CONFIG_BT_CTLR_DF=y
+CONFIG_BT_DF_CONNECTIONLESS_CTE_RX=y
+CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=n
```
Created nrf52833df_nrf52833.conf.
Will it be ok for you to align direction_finding_connectionelss_tx to this sample in the PR?

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,368 @@
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
```
Removed

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,368 @@
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/direction.h>
+#include <host/hci_core.h>
```
It was not needed here. I've removed it.

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -38,6 +38,9 @@
 #include "lll/lll_adv_pdu.h"
 #include "ll_sw/lll_sync_iso.h"
 #include "ll_sw/lll_scan.h"
+#if defined(CONFIG_BT_CTLR_DF)
+#include "lll/lll_df_types.h"
+#endif /* CONFIG_BT_CTLR_DF */
```
Added lll_df_types.h to OpenISA LLL.

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -48,6 +51,9 @@
 #include "ll_sw/ull_conn_types.h"
 #include "ll_sw/ull_conn_internal.h"
 #include "ll_sw/ull_conn_iso_types.h"
+#if defined(CONFIG_BT_CTLR_DF)
+#include "ll_sw/ull_df.h"
+#endif /* CONFIG_BT_CTLR_DF */
```
Conditions around include removed.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -34,3 +34,67 @@ struct lll_df_adv_cfg {
+	int16_t q;
+};
+
+/* Type defined on purpose, to give possibility to change maximum number
+ * of samples for proprietary implementations.
+ */
+typedef uint8_t sample_count_t;
```
Removed

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -34,3 +34,67 @@ struct lll_df_adv_cfg {
+/* Receive node aimed to report collected IQ samples during CTE receive */
+struct node_rx_iq_report {
+	struct node_rx_hdr hdr;
+	sample_count_t sample_count;
```
Changed to uint8_t.

ppryga commented:
@cvinayak All changes requested were implemented.

ppryga commented:
Fixed issue with coding rule - 5.7 A tag name shall be a unique identifier.
Referring to failing rules 5.1 and 5.2 I've created an issue: https://github.com/zephyrproject-rtos/zephyr/issues/34823

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	int16_t rssi;
+	/** Id of antenna used to measure the RSSI. */
+	uint8_t rssi_ant_id;
+	/** Type of CTE:
```
Would it make sense to add an enum for these values?
same for `slot_duration` and `packet_status`.

It would be better to have definitions in this file instead of the application using definitions in hci.h IMO.

ppryga replied:
Yes I can create enums in direction.h for `cte_type`, `slot_duration` and `packet_status`. I would have done that already but I didn't want to duplicate the enums. Those in hci.h are used by controller, so they have to stay there.

What is your prefered way?

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -59,4 +140,48 @@ int bt_df_adv_cte_tx_enable(struct bt_le_ext_adv *adv);
+ * Adds the callback structure to the list of callback structures that monitors
+ * CTE receive and sampling for periodic advertising sync.
+ *
+ * @param cb                 Callback struct. Must point to memory that remains
```
Aligning against a different comment block is not needed. I would remove it, it looks better without.

ppryga replied:
I'll correct it.

joerchan commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+
+struct bt_df_per_adv_sync_cte_report_cb {
+	/**
+	 * @brief Callback for IQ samples report collected when sampling
```
Is IQ sampling something everyone knows what means? (I don't).
Is it explained somewhere?

ppryga replied:
IQ samples are not explained in code documentation anywhere. I assumed that one who wants to use Direction Finding has already gained appropriate information.

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -189,6 +200,151 @@ static int hci_df_set_adv_cte_tx_enable(struct bt_le_ext_adv *adv,
+		}
+
+		if (params->cte_type & BT_DF_CTE_RX_TYPE_AOA) {
+			if (df_ant_info.num_ant <
+				    DF_SAMPLING_ANTENNA_NUMBER_MIN ||
+			    !BT_FEAT_LE_ANT_SWITCH_RX_AOA(bt_dev.le.features)) {
```
Minor:
```suggestion
			if (df_ant_info.num_ant < DF_SAMPLING_ANTENNA_NUMBER_MIN ||
			    !BT_FEAT_LE_ANT_SWITCH_RX_AOA(bt_dev.le.features)) {
```
To make it significantly more readable without crossing the 100 char rule.

ppryga replied:
Updates, unfortunately .clang-format has 80 characters limit still. That is the cause for strange formatting in the PR.

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -189,6 +200,151 @@ static int hci_df_set_adv_cte_tx_enable(struct bt_le_ext_adv *adv,
 }
 
+#if defined(CONFIG_BT_DF_CONNECTIONLESS_CTE_RX)
+static int hci_df_set_cl_cte_rx_enable(struct bt_le_per_adv_sync *sync, bool enable,
+				       const struct bt_df_per_adv_sync_cte_rx_param *params)
+{
```
With the several layers of indentation, as well as being 100 lines long, I think perhaps some of the code could be moved into intermediary static functions to significantly improve readability, e.g. the first `if (enable)` clause seems to be a validation of the parameters.

ppryga replied:
Extracted two additional functions.

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -189,6 +200,151 @@ static int hci_df_set_adv_cte_tx_enable(struct bt_le_ext_adv *adv,
+	}
+
+	if (!(per_adv_sync->cte_type & BIT(evt->cte_type))) {
+		BT_DBG("CTE filtered out by cte_type:%u", evt->cte_type);
```
```suggestion
		BT_DBG("CTE filtered out by cte_type: %u", evt->cte_type);
```

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -348,3 +504,49 @@ int bt_df_adv_cte_tx_disable(struct bt_le_ext_adv *adv)
+int bt_df_per_adv_sync_cte_rx_enable(struct bt_le_per_adv_sync *sync,
+				     const struct bt_df_per_adv_sync_cte_rx_param *params)
+{
+	__ASSERT_NO_MSG(sync);
+	__ASSERT_NO_MSG(params);
```
We might want to consider using the `CHECKIF` macro here instead, and simply return an error (that way the behavior of sending invalid pointers can be configured)

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -348,3 +504,49 @@ int bt_df_adv_cte_tx_disable(struct bt_le_ext_adv *adv)
+	return bt_df_set_per_adv_sync_cte_rx_enable(sync, false, NULL);
+}
+
+void bt_df_per_adv_sync_cte_cb_register(struct bt_df_per_adv_sync_cte_report_cb *cb)
+{
+	sys_slist_append(&df_per_adv_sync_cbs, &cb->node);
+}
+
+void bt_df_per_adv_sync_cte_cb_unregister(struct bt_df_per_adv_sync_cte_report_cb *cb)
+{
+	sys_slist_find_and_remove(&df_per_adv_sync_cbs, &cb->node);
+}
```
Since you are already asserting the pointers above, perhaps also a check for `cb` here (as you do `cb->node`).

ppryga replied:
These functions were removed because callback were added to PS sync callbacks structure.

Thalley commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -159,6 +159,11 @@ enum {
 	/** Periodic advertising is attempting sync sync */
 	BT_PER_ADV_SYNC_RECV_DISABLED,
 
+	/* Constant Tone Extension for Periodic Advertising has been enabled
```
```suggestion
	/** Constant Tone Extension for Periodic Advertising has been enabled
```
Not really needed, but would make it more consistent

ppryga replied:
Corrected

Thalley commented on tests/bluetooth/df/common/src/radio_df_stub.c:
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2021 Nordic Semiconductor ASA
```
or
```suggestion
 * Copyright (c) 2020-2021 Nordic Semiconductor ASA
```

ppryga replied:
Corrected

Thalley commented on tests/bluetooth/df/connectionless_cte_rx/src/test_set_iq_sampling_enable.c:
```diff
@@ -0,0 +1,341 @@
+ * @param[in]sync_handle                 Handle of sync set.
+ * @param[in]sync_flags                  Flags related with sync set.
+ * @param[in]params                      CTE Sampling parameters.
+ * @param[in] enable                     Enable or disable CTE RX
```
```suggestion
 * @param[in]enable                     Enable or disable CTE RX
```

ppryga replied:
Corrected.

Thalley commented on tests/bluetooth/df/connectionless_cte_rx/src/test_set_iq_sampling_enable.c:
```diff
@@ -0,0 +1,341 @@
+/* @brief Function sends HCI_LE_Set_Connectionless_CTE_Sampling_Enable
+ *        to controller.
+ *
+ * @param[in]sync_handle                 Handle of sync set.
+ * @param[in]sync_flags                  Flags related with sync set.
+ * @param[in]params                      CTE Sampling parameters.
+ * @param[in] enable                     Enable or disable CTE RX
```
The `[in]` seems like overkill here as well

ppryga replied:
Removed.

Thalley commented on tests/bluetooth/df/connectionless_cte_rx/src/test_set_iq_sampling_enable.c:
```diff
@@ -0,0 +1,341 @@
+			uint8_t *dest_ant_ids =
+				net_buf_add(buf, params->num_ant_ids);
+
+			memcpy(dest_ant_ids, params->ant_ids,
```
For some reason we much more often do `(void)memset` than `(void)memcpy`, while we should probably do latter if we do the former :D

Thalley commented on tests/bluetooth/df/connectionless_cte_tx/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
```
```suggestion
# Copyright (c) 2021 Nordic Semiconductor ASA
```

ppryga replied:
Corrected

Thalley commented on tests/bluetooth/df/connectionless_cte_tx/src/common.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021 Nordic Semiconductor ASA
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Why was this changed to 2020? Should it perhaps be 2020-2021?

ppryga replied:
Corrected

Thalley commented on tests/bluetooth/df/connectionless_cte_tx/src/common.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021 Nordic Semiconductor ASA
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Why was this changed to 2020? Should it perhaps be 2020-2021?

ppryga replied:
Corrected

Thalley reviewed (changes requested):
I skipped the controller part, as I am sure there are more qualified people than me that can review that part. 

Mostly small comments, but otherwise a very throughout implementation so far.

jhedberg commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,366 @@
+	printk("Periodic Advertising CTE receive callbacks register...");
+	bt_df_per_adv_sync_cte_cb_register(&cte_report_callbacks);
+	printk("success.\n");
+	return 0;
```
I probably did, but it's one of the conventions I don't feel that strongly about. Here however it does make the code look slightly nicer looking (IMO) since you've then got the three similar blocks more clearly separated.

ppryga replied:
Empty line added.

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
yes, it should. Corrected. Also corrected in sample direction_finding_connectionless_tx/CMakelists.txt.

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,366 @@
+		    const struct bt_le_per_adv_sync_term_info *info)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	int64_t time_delta_ms;
+
+	time_delta_ms = k_uptime_get() - start_time;
```
It was leftover from testing. Removed.

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,366 @@
+	k_sem_give(&sem_per_sync_lost);
+}
+
+bool enable = true;
```
It was leftover from testing. Removed.

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,366 @@
+	if (!scan_enabled) {
+		printk("Start scanning...");
+		err = bt_le_scan_start(&param, NULL);
+		if (err) {
```
I'll correct it at least in the sample application code.

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/issues/34832 :)

ppryga commented on samples/bluetooth/direction_finding_connectionless_rx/src/main.c:
```diff
@@ -0,0 +1,366 @@
+		}
+		printk("success\n");
+		scan_enabled = true;
+	}
+	return 0;
```
Empty line added

ppryga-nordic committed:
Bluetooth: controller: Use common mem_link_rx for node_rx_iq_report

Change use of dedicated memory pool for linked list nodes for
node_rx_iq_report to common mem_link_rx. Former solution had
a drawback. Released link nodes may be enqueued to wrong memory
pool. E.g. link related with nopde_rx_iq_report went to common
link memory pool, whereas link nodes from common pool were enqueued
to dedicated list.
The solution was working because links have the same memory layout,
just different memory pools they originated from.
The problem may occur if one of those link memory pools is reset.
Then the same link may be used by node_rx and node_rx_iq_report
at the same time, causing controller failure.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	(BT_DF_CTE_RX_TYPE_AOA | BT_DF_CTE_RX_TYPE_AOD_1US |                   \
+	 BT_DF_CTE_RX_TYPE_AOD_2US)
+
+/* @brief Constant Tone Extension parameters for connectionless
```
I'll go with first one to have the same formatting in whole file. I'll update it later in a separate PR with appropriate explanation.

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+ * for correctness.
+ */
+struct bt_df_per_adv_sync_cte_rx_param {
+	/* Bitmap with allowed CTE types: AoD 1us, AoD 2us, AoA */
```
Ok

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	uint8_t slot_durations;
+	/** Max number of CTEs to receive, 0 means receive continuously. */
+	uint8_t max_cte_count;
+	/** Length of antenna switch pattern */
```
Missing dots added.

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -59,4 +140,48 @@ int bt_df_adv_cte_tx_enable(struct bt_le_ext_adv *adv);
+ * Receive and sampling of Constant Tone Extension may be enabled only after
+ * periodic advertising sync is established.
+ *
+ * @param[in] sync           Periodic advertising sync object.
+ * @param[in] params         CTE receive and sampling parameters.
```
I've removed [in].

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	int16_t rssi;
+	/** Id of antenna used to measure the RSSI. */
+	uint8_t rssi_ant_id;
+	/** Type of CTE:
```
Added reference to created enum. The comment is single line now.

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	 * - 0x2 Angle of departure with 2us slots
+	 */
+	uint8_t cte_type;
+	/** Duration of slots when received CTE type is AoA:
```
Added reference to created enum. The comment is single line now.

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -26,6 +26,87 @@ struct bt_df_adv_cte_tx_param {
+	 * - 0x2 2us slots
+	 */
+	uint8_t slot_durations;
+	/** Status of received PDU with CTE:
```
Added reference to created enum. The comment is single line now.

ppryga commented on include/bluetooth/direction.h:
```diff
@@ -59,4 +140,48 @@ int bt_df_adv_cte_tx_enable(struct bt_le_ext_adv *adv);
+ * Removes the callback structure to the list of callback structures that
+ * monitors CTE receive and sampling for periodic advertising sync.
+ *
+ * @param cb                 Callback struct. Must point to memory that remains
+ * valid.
```
These functions were removed because callback were added to PS sync callbacks structure.

ppryga-nordic committed:
bluetooth: Fix cte_type value in periodic adv sync report

Periodic advertising sync report returned wrong cte_type value.
This commit fixes the problem.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
All requested changes are done.
- updated commits related with: host, sample application, unit tests
- added commit to use common mem_link_rx pool instead of dedicated one for node_rx_iq_report
- added commit that fixes wrong cte_type in periodic advertising reports.

Coding guidelines still fail.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2589,6 +2590,7 @@ static void le_df_connectionless_iq_report(struct pdu_data *pdu_rx,
 	}
 
 	sep->packet_status = iq_report->packet_status;
+	printk("packet status %u\n", sep->packet_status);
```
Remove printk

cvinayak reviewed (changes requested):
Minor correction to commit message "Released link nodes may be enqueued to wrong memor"  missing "y". ~And remove a printk.~

ppryga commented:
@cvinayak Commit message corrected as well as: removed printk and fixed code formatting in ull_df_internal.h.
I had to update commit that fixes builds for non Nordic platforms. There were missing forward `struct lll_df_sync` declaration. With that unit tests for OpenISA should pass.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -7,27 +7,108 @@
+	BT_DF_CTE_TYPE_NONE = 0,
+	/** Angle of Arrival mode. Antenna switching done on receiver site. */
+	BT_DF_CTE_TYPE_AOA = BIT(0),
+	/** Angle of Departure mode with 1 us antenna switching slots.
+	 *  Antenna switching done on transmitter site.
+	 */
```
```suggestion
	/** 
	 * @brief Angle of Departure mode with 1 us antenna switching slots.
	 * 
	 *  Antenna switching done on transmitter site.
	 */
```

ppryga replied:
Corrected according to suggestion.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -7,27 +7,108 @@
+	 *  Antenna switching done on transmitter site.
+	 */
+	BT_DF_CTE_TYPE_AOD_1US = BIT(1),
+	/** Angle of Departure mode with 2 us antenna switching slots.
+	 *  Antenna switching done on transmitter site.
+	 */
```
```suggestion
	/** 
	 * @brief Angle of Departure mode with 2 us antenna switching slots.
	 * 
	 *  Antenna switching done on transmitter site.
	 */
```

ppryga replied:
Corrected according to suggestion.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -7,27 +7,108 @@
+enum bt_df_packet_status {
+	/** Received PDU had CRC OK */
+	BT_DF_CTE_CRC_OK = 0x0,
+	/** Received PDU had incorrect CRC, but Radio peripheral
+	 * was able to parse CTEInfo field of the PDU and process
+	 * sampling of CTE.
+	 */
```
```suggestion
	/** 
	 * @brief Received PDU had incorrect CRC
	 * 
	 * Incorrect CRC but Radio peripheral
	 * was able to parse CTEInfo field of the PDU and process
	 * sampling of CTE.
	 */
```

ppryga replied:
Corrected according to suggestion.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -7,27 +7,108 @@
+	 * sampling of CTE.
+	 */
+	BT_DF_CTE_CRC_ERR_CTE_BASED_TIME = 0x1,
+	/** Received PDU had incorrect CRC, but Radio peripheral
+	 * was able to process sampling of CTE in some other way.
+	 */
```
```suggestion
	/** 
	 * @brief Received PDU had incorrect CRC
	 * 
	 * Incorrect CRC but Radio peripheral
	 * was able to process sampling of CTE in some other way.
	 */
```

ppryga replied:
Corrected according to suggestion.

Thalley commented on samples/bluetooth/direction_finding_connectionless_rx/README.rst:
```diff
@@ -0,0 +1,78 @@
+  :zephyr_file:`samples/bluetooth/direction_finding_connectionless_rx/boards/nrf52833dk_nrf52833.overlay`
+  file
+* Update the antenna switching patterns in :cpp:var:`ant_patterns` array in
+  :zephyr_file:`samples/bluetooth/direction_finding_connectionless_tx/src/main.c`..
```
```suggestion
  :zephyr_file:`samples/bluetooth/direction_finding_connectionless_tx/src/main.c`.
```

ppryga replied:
Dot removed.

Thalley commented on samples/bluetooth/direction_finding_connectionless_rx/boards/nrf52833dk_nrf52833.overlay:
```diff
@@ -0,0 +1,26 @@
+	/* This is an example number of antennas that may be available
+	 * on antenna matrix board.
+	 */
+	dfe-antenna-num = < 10 >;
+	/* This is an example switch pattern that will be used to set an
+	 * antenna for Tx PDU (period before start of Tx CTE).
+	 */
+	dfe-pdu-antenna = <0x1>;
```
Minor: One place you have spaces between the value and the brackets (< >), and later you don't.

ppryga replied:
Spaces removed.

Thalley reviewed (changes requested):
A few very minor things that should be fixed, otherwise LGTM

ppryga commented:
@Thalley latest suggestions implemented.

Thalley reviewed (approved):
LGTM
